### ISSUE TITLE
UnsupportedFeatureError when applying schema where a constraint contains a potentially empty argument.

### ISSUE THREAD

**Issue by HamishWHC** [edgedb/edgedb#4926]
- EdgeDB Version: 2.9+4fbadd2
- EdgeDB CLI Version: EdgeDB CLI 2.2.6+7eabbf9
- OS Version: macOS 12.4

Steps to Reproduce:

1. Create default module as follows:
```
module default {
    type Demo {
        required property start_time -> datetime {
            default := now();
            readonly := true;
        };
        property end_time -> datetime;
        constraint expression on ((not exists .end_time) or (.start_time < .end_time));
    }
}
```
2. Run `edgedb migration create` and `edgedb migrate`. An error occurs during the `migrate` command: `edgedb error: UnsupportedFeatureError: operations on potentially empty arguments not supported in simple expressions`

The issue is that `.start_time < .end_time` may be empty (since `end_time` is not required) and `or` would potentially have an empty argument. This can be fixed with a `?? false`.

The bug is that this is not caught by the `create` command and that such an unhelpful error is produced. I had to slowly comment out sections of my schema and reset the DB to find the issue. I've seen `create` catch these kinds of errors before, so this was rather frustrating.

Not in the scope of the bug, but more as a discussion point: why not have boolean operators (and `expression` constraints) treat `{}` as falsy?

**Comment by msullivan** [comment]
There's a surprisingly large amount to unpack here:
1. I agree that it is a bug that we report this at `migrate` time instead of at `create migration` time. It is also especially a bug that we don't report error locations. The root cause of this is that this error (and several more like it) are only caught by the IR->SQL compiler, which only runs when actually applying the migration.
2. I think it is also actually a bug that `or`/`and` (and a number of other functions) don't work in simple expressions on potentially empty sets. The reason for this is that the SQL operators we call for `OR` and `AND` (and others) are not "strict"--they sometimes return a non-NULL result on NULL inputs. This does not match EdgeQL semantics, so when we use those operations on potentially NULL values, we need to insert WHERE clauses that filter out NULLs. Unfortunately check constraints and many other places do not allow subqueries, so we can't put WHERE clauses, so we can't properly compile these operations. This is fixable, though. (The easiest approach would probably be to have a two-argument strict function that always returns its first argument, and insert calls to that to filter).
3. That bug is arguably a blessing in disguise, though, because expression constraints treat `{}` as being *truthy*. This is behavior inherited directly from postgres, but I don't love that behavior. The best option might be to reject possibly empty constraint expressions, but we'd have to have a migration path. (Though constraints *on* an empty property should still only fail if it is set?)
4. The core semantics of edgeql is that using an empty set as the argument to a non-optional parameter produces an empty set, and we wanted optional handling functions to be explicit: `??`, `?=`, `?!=` all make it pretty clear that something special is happening with the `?`.

**Comment by msullivan** [comment]
This was fixed in #5271.
Though we still treat `{}` as truthy, via postgres, so it's probably still best to avoid this.
